sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,707 @@
+configuration file that has one entry for each device that it supports.  Each
+entry contains the set of aspects that the device supports and the set of
+sub-group sizes that it supports.  DPC++ then consults this configuration
+file to decide whether to invoke a particular AOT compiler on each device IR
+module, using the information from the module's "SYCL/image-requirements"
+property set.
+
+#### Device configuration file
+
+The configuration file uses a simple YAML format where each top-level key is
+a name of a device architecture.  We expect to define a set of device
+architecture names that are used consistently in many places (in this
+configuration file, in the names of device-specific aspects, as parameters for
+the `-fsycl-targets` command line option, etc.)  However, we have not yet
+agreed on these architecture names.  There are sub-keys under each device for
+the supported aspects, sub-group sizes and AOT compiler ID.  For example:
+
+```
+gen11_1:
+  aspects: [1, 2, 3]
+  sub-group-sizes: [8, 16]
+gen_icl:
+  aspects: [2, 3]
+  sub-group-sizes: [8, 16]
+x86_64_avx512:
+  aspects: [1, 2, 3, 9, 11]
+  sub-group-sizes: [8, 32]
+```
+
+The values of the aspects in this configuration file can be the numerical
+values from the `enum class aspect` enumeration or the enum identifier itself.
+
+One advantage to encoding this information in a textual configuration file is
+that customers can update the file if necessary.  This could be useful, for
+example, if a new device is released before there is a new DPC++ release.  In
+fact, the DPC++ driver supports a command line option which allows the user
+to select an alternate configuration file.
+
+**TODO**: More information will be inserted here when we merge
+[this separate PR][6] into this design document.
+
+[6]: <https://github.com/gmlueck/llvm/pull/1>
+
+
+### Changes to the DPC++ runtime
+
+The DPC++ runtime must be changed to check if a kernel uses any optional
+features that the device does not support.  If this happens, the runtime must
+raise a synchronous `errc::kernel_not_supported` exception.
+
+When the application submits a kernel to a device, the runtime identifies all
+the other device images that export device functions which are needed by the
+kernel as described in [Device Code Dynamic Linking][4].  Before the runtime
+actually links these images together, it compares each image's
+"SYCL/image-requirements" against the features provided by the target
+device.  If any of the following checks fail, the runtime throws
+`errc::kernel_not_supported`:
+
+* The "aspect" property contains an aspect that is not provided by the device,
+  or
+* The "reqd\_sub\_group\_size" property contains a sub-group size that the
+  device does not support.
+
+There is no way currently for the runtime to query the work-group sizes that a
+device supports, so the "reqd\_work\_group\_size" property is not checked.  We
+include this property in the set nonetheless for possible future use.
```

Comment:
There are a lot of exceptions in this document about `reqd_work_group_size` (and friend).  Maybe it simply isn't a good match for being an optional kernel feature.  Perhaps it might be best to _not_ include it now. Wouldn't that simplify things?

Reply:
[Section 5.7. "Optional kernel features"](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features) of the SYCL 2020 spec lists both `reqd_work_group_size` and `reqd_sub_group_size` as optional kernel features, so I think the design needs to account for them.